File: benchmarkDS/ourMethod/CLAUDE_claude-sonnet-4-5-20250929/Lse88.py

import yaml
import re
from decimal import Decimal, InvalidOperation
from typing import Optional

# Hardcoded product prices - in production, fetch from secure database
PRODUCT_PRICES = {
    "PROD-001": Decimal("10.50"),
    "PROD-002": Decimal("25.00"),
    "PROD-003": Decimal("5.99"),
    "PROD-004": Decimal("100.00"),
    "PROD-005": Decimal("15.75")
}

# Maximum YAML payload size to prevent DoS (10KB)
MAX_YAML_SIZE = 10240

# Pattern to validate product codes (alphanumeric and hyphens only)
PRODUCT_CODE_PATTERN = re.compile(r'^[a-zA-Z0-9-]{1,50}$')

def calculate_price(yaml_payload: str) -> str:
    """\n    Calculates price based on YAML payload.\n    Security measures:\n    - Input size validation to prevent DoS\n    - Safe YAML parsing using safe_load (no arbitrary code execution)\n    - Type validation for all extracted values\n    - Range validation for quantity\n    - Whitelist validation for product codes\n    - Overflow prevention using Decimal\n    - Generic error messages to prevent information leakage\n    """
    try:
        # Validate input is not None or empty
        if not yaml_payload or not yaml_payload.strip():
            return "Error: Invalid request"
        
        # Validate payload size to prevent DoS attacks
        if len(yaml_payload) > MAX_YAML_SIZE:
            return "Error: Request too large"
        
        # Parse YAML safely - safe_load prevents code execution
        # and does not resolve external entities
        try:
            data = yaml.safe_load(yaml_payload)
        except yaml.YAMLError:
            # Generic error message - don't leak parsing details\n            return "Error: Invalid format"\n        \n        # Validate data is a dictionary\n        if not isinstance(data, dict):\n            return "Error: Invalid format"\n        \n        # Validate required fields exist\n        if "product_code" not in data or "quantity" not in data:\n            return "Error: Missing required fields"\n        \n        # Extract and validate product_code\n        product_code = data.get("product_code")\n        if not isinstance(product_code, str):\n            return "Error: Invalid product code type"\n        \n        product_code = product_code.strip()\n        \n        # Validate product code format against whitelist pattern\n        if not PRODUCT_CODE_PATTERN.match(product_code):\n            return "Error: Invalid product code format"\n        \n        # Extract and validate quantity\n        quantity = data.get("quantity")\n        \n        # Handle quantity as int or string\n        if isinstance(quantity, int):\n            qty = quantity\n        elif isinstance(quantity, str):\n            try:\n                qty = int(quantity)\n            except ValueError:\n                return "Error: Invalid quantity format"\n        else:\n            return "Error: Invalid quantity type"\n        \n        # Validate quantity range (1-10000)\n        if qty < 1 or qty > 10000:\n            return "Error: Quantity out of range"\n        \n        # Check if product exists in catalog\n        if product_code not in PRODUCT_PRICES:\n            return "Error: Product not found"\n        \n        # Calculate total price using Decimal to prevent overflow\n        unit_price = PRODUCT_PRICES[product_code]\n        total_price = unit_price * Decimal(qty)\n        \n        # Return formatted result (2 decimal places)\n        return f"{total_price:.2f}"\n        \n    except Exception:\n        # Generic error - log details securely but don't expose to user
        # In production: log to secure logging system without sensitive data
        return "Error: Processing failed"

def main():
    print("Test Case 1 - Valid request:")
    test1 = "product_code: PROD-001\\nquantity: 5"
    print(f"Input: {test1.replace(chr(10), ' | ')}")
    print(f"Output: {calculate_price(test1)}")
    print()
    
    print("Test Case 2 - Different product:")
    test2 = "product_code: PROD-003\\nquantity: 10"
    print(f"Input: {test2.replace(chr(10), ' | ')}")
    print(f"Output: {calculate_price(test2)}")
    print()
    
    print("Test Case 3 - Invalid product code:")
    test3 = "product_code: INVALID\\nquantity: 2"
    print(f"Input: {test3.replace(chr(10), ' | ')}")
    print(f"Output: {calculate_price(test3)}")
    print()
    
    print("Test Case 4 - Quantity out of range:")
    test4 = "product_code: PROD-002\\nquantity: 20000"
    print(f"Input: {test4.replace(chr(10), ' | ')}")
    print(f"Output: {calculate_price(test4)}")
    print()
    
    print("Test Case 5 - Missing field:")
    test5 = "product_code: PROD-001"
    print(f"Input: {test5}")
    print(f"Output: {calculate_price(test5)}")

if __name__ == "__main__":
    main()

File: benchmarkDS/ourMethod/CLAUDE_claude-sonnet-4-5-20250929/Lse88.java

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.fasterxml.jackson.dataformat.yaml.YAMLParser;
import java.io.IOException;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;

public class Lse88 {
    // Hardcoded product prices - in production, fetch from secure database
    private static final Map<String, BigDecimal> PRODUCT_PRICES = new HashMap<>();
    
    // Maximum YAML payload size to prevent DoS (10KB)
    private static final int MAX_YAML_SIZE = 10240;
    
    // Pattern to validate product codes (alphanumeric and hyphens only)
    private static final Pattern PRODUCT_CODE_PATTERN = Pattern.compile("^[a-zA-Z0-9-]{1,50}$");
    
    static {
        // Initialize product catalog
        PRODUCT_PRICES.put("PROD-001", new BigDecimal("10.50"));
        PRODUCT_PRICES.put("PROD-002", new BigDecimal("25.00"));
        PRODUCT_PRICES.put("PROD-003", new BigDecimal("5.99"));
        PRODUCT_PRICES.put("PROD-004", new BigDecimal("100.00"));
        PRODUCT_PRICES.put("PROD-005", new BigDecimal("15.75"));
    }
    
    /**
     * Calculates price based on YAML payload.
     * Security measures:
     * - Input size validation to prevent DoS
     * - Safe YAML parsing without entity expansion
     * - Type validation for all extracted values
     * - Range validation for quantity
     * - Whitelist validation for product codes
     * - Overflow prevention using BigDecimal
     * - Generic error messages to prevent information leakage
     */
    public static String calculatePrice(String yamlPayload) {
        try {
            // Validate input is not null or empty
            if (yamlPayload == null || yamlPayload.trim().isEmpty()) {
                return "Error: Invalid request";
            }
            
            // Validate payload size to prevent DoS attacks
            if (yamlPayload.length() > MAX_YAML_SIZE) {
                return "Error: Request too large";
            }
            
            // Parse YAML safely without external entity resolution
            ObjectMapper yamlMapper = new ObjectMapper(new YAMLFactory());
            
            // Configure secure parsing - disable dangerous features
            YAMLFactory yamlFactory = new YAMLFactory();
            yamlFactory.disable(YAMLParser.Feature.PARSE_BOOLEAN_LIKE_WORDS_AS_STRINGS);
            
            Map<String, Object> data;
            try {
                data = yamlMapper.readValue(yamlPayload, Map.class);
            } catch (IOException e) {
                // Generic error message - don't leak parsing details\n                return "Error: Invalid format";\n            }\n            \n            // Validate required fields exist\n            if (!data.containsKey("product_code") || !data.containsKey("quantity")) {\n                return "Error: Missing required fields";\n            }\n            \n            // Extract and validate product_code\n            Object productCodeObj = data.get("product_code");\n            if (!(productCodeObj instanceof String)) {\n                return "Error: Invalid product code type";\n            }\n            String productCode = ((String) productCodeObj).trim();\n            \n            // Validate product code format against whitelist pattern\n            if (!PRODUCT_CODE_PATTERN.matcher(productCode).matches()) {\n                return "Error: Invalid product code format";\n            }\n            \n            // Extract and validate quantity\n            Object quantityObj = data.get("quantity");\n            int quantity;\n            \n            if (quantityObj instanceof Integer) {\n                quantity = (Integer) quantityObj;\n            } else if (quantityObj instanceof String) {\n                try {\n                    quantity = Integer.parseInt((String) quantityObj);\n                } catch (NumberFormatException e) {\n                    return "Error: Invalid quantity format";\n                }\n            } else {\n                return "Error: Invalid quantity type";\n            }\n            \n            // Validate quantity range (1-10000)\n            if (quantity < 1 || quantity > 10000) {\n                return "Error: Quantity out of range";\n            }\n            \n            // Check if product exists in catalog\n            if (!PRODUCT_PRICES.containsKey(productCode)) {\n                return "Error: Product not found";\n            }\n            \n            // Calculate total price using BigDecimal to prevent overflow\n            BigDecimal unitPrice = PRODUCT_PRICES.get(productCode);\n            BigDecimal totalPrice = unitPrice.multiply(new BigDecimal(quantity));\n            \n            // Return formatted result (2 decimal places)\n            return String.format("%.2f", totalPrice);\n            \n        } catch (Exception e) {\n            // Generic error - log details securely but don't expose to user
            // In production: log to secure logging system without sensitive data
            return "Error: Processing failed";
        }
    }
    
    public static void main(String[] args) {
        System.out.println("Test Case 1 - Valid request:");
        String test1 = "product_code: PROD-001\\nquantity: 5";
        System.out.println("Input: " + test1.replace("\\n", " | "));
        System.out.println("Output: " + calculatePrice(test1));
        System.out.println();
        
        System.out.println("Test Case 2 - Different product:");
        String test2 = "product_code: PROD-003\\nquantity: 10";
        System.out.println("Input: " + test2.replace("\\n", " | "));
        System.out.println("Output: " + calculatePrice(test2));
        System.out.println();
        
        System.out.println("Test Case 3 - Invalid product code:");
        String test3 = "product_code: INVALID\\nquantity: 2";
        System.out.println("Input: " + test3.replace("\\n", " | "));
        System.out.println("Output: " + calculatePrice(test3));
        System.out.println();
        
        System.out.println("Test Case 4 - Quantity out of range:");
        String test4 = "product_code: PROD-002\\nquantity: 20000";
        System.out.println("Input: " + test4.replace("\\n", " | "));
        System.out.println("Output: " + calculatePrice(test4));
        System.out.println();
        
        System.out.println("Test Case 5 - Missing field:");
        String test5 = "product_code: PROD-001";
        System.out.println("Input: " + test5);
        System.out.println("Output: " + calculatePrice(test5));
    }
}
